### ISSUE TITLE
[RFC] Route parameter constraints

### ISSUE THREAD

**Issue by Sayan751** [aurelia/aurelia#1734]
## TL;DR

- Proposal of route parameter constraint based on RegExp API and pattern. Example: `'product/:id{{\d+}}'`
- Usage is optional; when present, the route recognizer will respect that.
- Out-of-scope: The usage of wildcard pattern using `*`, `**` etc.
- Limitation: The usage of `/` in the pattern is not allowed.

# 💬 RFC

This RFC proposes application of pattern constraints for route parameter.

## 🔦 Context

The route parameters can be required, optional, or wildcard. So far, there is no other kind of constrains applied on the parameters. However, sometimes it might be desirable to restrict the navigation only if the given parameters meet certain patterns. This RFC proposes an extension in the route parameters recognition based on optional regex pattern. If no pattern is specified, then the route recognizer works as it works now. However, if a constraint is specified for a route parameter, then the route recognizer only matches/recognizes the given route, if and only if all the constraints specified for all the route parameters are fulfilled.

This is explained further using the examples below.

## 💻 Examples

Restrict the `id` parameter to the `product` route only to integer.

```typescript
'product/:id{{\d+}}'
```

The `{{` and `}}` are the **terminal tokens** that indicate the start and the end of the pattern. AFAIK, that should not pose a conflict to the possible regex syntaxes. The [URLPattern API](https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API) uses `(`, and `)` as terminal tokens. However, I think that adopting a non-conflicting token will keep the parsing logic easy and also make scope for more intricate regex stuff in future; for example, capture groups.

Let us consider another example. The following example, restricts the `tab` parameter to 'items', 'payment', and 'delivery'.

```typescript
'order/:tab?{{items|payment|delivery}}'
```

Note that we don't want to parse the regex by ourselves, and rather use the **standard RegExp API** for that. Hence, what is supported by the RegExp API should also be supported by this.

This also means that glob/wildcard patterns like below will not be supported and are **out of the scope** of the change proposed by this RFC.

```typescript
'view-file/:name?{{*.txt}}' // out of the scope
```

However, one can easily write a regex for that; a naive approach may look like below.

```typescript
'view-file/:name?{{^.+\.txt$}}'
```

**One limitation** is the usage of `/` in the parameter constraint. Because in the context of URL, that is a path separator and a parameter value can never contain that.

## Related works

- [ASP.NET Web API route constraints](https://learn.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/attribute-routing-in-web-api-2#route-constraints).
- [URLPattern API](https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API).

## Future extension possibilities (currently out of the scope of this proposal)

- A more generic syntax: `'product/:id{{$contraint:regex:\d+}}'`. 
- Out of the box patterns: `{{$constraint:int}}`, `{{$constraint:range:10:42}}` etc.
- Custom constraint: `{{$constraint:notInBlacklist:foo:bar}}`.
- `IConstraint`

  ```ts
  interface IConstraint<T = unknown> {
    hydrate(raw: string): void;
    match(raw: string): boolean;
    coerce(raw: string): T;
  }
  ```
- Constraints metadata can be registered and resolved, enabling constraint discovery by name: `@routeConstriant('notInBlacklist')`.
- Type coercion according to the constraints.

**Comment by bigopon** [comment]
I think the RFC needs to mention
- prior arts
- use case/examples
- next steps, just so we are clear whether this can allow expansion of feature. Example: an expansion on the syntaxes, to allow custom "resolver" `/:tab{!{MyTabs}!}` etc...

Also, will the router and the router-lite behave similarly once we start introducing this? @jwx

**Comment by Sayan751** [comment]
@bigopon I have updated the RFC. Note that the same route-recognizer is shared by both routers. As long as it holds, this will be automatically available to both routers.

**Comment by mroeling** [comment]
Few questions/remarks if I may. 
1) How does this functionality compare to other frameworks? 
2) Should it be the task of the router to apply limitations/checks, or should that be performed by the class that retrieves/handles the route? Personally I would prefer the latter.

**Comment by Sayan751** [comment]
@mroeling 

> 1) How does this functionality compare to other frameworks? 

Personally I have only experience with ASP.NET Web API (refer the link posted in the original post). The support there is pretty broad with custom constraints etc.. I think that is a good target to achieve. For the first iteration I think it is enough to support regex. And then in future iterations we can gradually work on other extensions of this feature. 

Similar feature exists also in URLPattern API and react-router. AFAIK both of those inherit the concepts from path-to-regex.

> 2) Should it be the task of the router to apply limitations/checks, or should that be performed by the class that retrieves/handles the route? Personally I would prefer the latter.

I don't think that I understand this question clearly. As I envision it, the classes that define the routes can specify/apply the constraints for the paths. 
The route recognizer will be responsible to parse the paths including the constraints. It will also be responsible to respect the constraints while  recognizing/matching the routes. For example, given the path `product/:id{{\d+}}`, the recognizer matches the path `product/42` but not `product/foo`.
If we delegate this responsibilities out of the scope of the router, then it will not be correctly encapsulated and will surely be overwhelming for the devs.

**Comment by glyad** [comment]
Great idea! Anyway, is the release and promotion of Aurelia 2 less significant, than the implementation of this idea?
